Userland/Applications/Mail/MailWidget.cpp

```diff
@@ -0,0 +1,480 @@
+
+        active_inbox_entries.append(inbox_entry);
+    }
+
+    m_individual_mailbox_view->set_model(InboxModel::create(active_inbox_entries));
+    m_individual_mailbox_view->on_selection_change = [this] {
+        selected_email_to_load();
+    };
+}
+
+void MailWidget::selected_email_to_load()
+{
+    auto const& index = m_individual_mailbox_view->selection().first();
+
+    if (!index.is_valid())
+        return;
+
+    // IMAP is 1-based.
+    int id_of_email_to_load = index.row() + 1;
+
+    auto fetch_command = IMAP::FetchCommand {
+        .sequence_set = { { id_of_email_to_load, id_of_email_to_load } },
+        .data_items = {
+            IMAP::FetchCommand::DataItem {
+                .type = IMAP::FetchCommand::DataItemType::BodyStructure,
+            },
+        },
+    };
+
+    auto fetch_response = m_imap_client->fetch(fetch_command, false)->await().release_value();
+
+    if (fetch_response.status() != IMAP::ResponseStatus::OK) {
+        dbgln("Failed to retrieve the body structure of the selected e-mail. The server says: '{}'", fetch_response.response_text());
+        GUI::MessageBox::show_error(window(), String::formatted("Failed to retrieve the selected e-mail. The server says: '{}'", fetch_response.response_text()));
+        return;
+    }
+
+    Vector<int> selected_alternative_position;
+    String selected_alternative_encoding;
+
+    auto& response_data = fetch_response.data().fetch_data().last().get<IMAP::FetchResponseData>();
+
+    response_data.body_structure().data().visit(
+        [&](IMAP::BodyStructureData const& data) {
+            // The message will be in the first position.
+            selected_alternative_position.append(1);
+            selected_alternative_encoding = data.encoding;
+        },
+        [&](IMAP::MultiPartBodyStructureData const& data) {
+            auto alternatives = get_alternatives(data);
+            if (alternatives.is_empty()) {
+                dbgln("No alternatives. The server said: '{}'", fetch_response.response_text());
+                GUI::MessageBox::show_error(window(), String::formatted("The server sent no message to display."));
+                return;
+            }
+
+            // We can choose whichever alternative we want. In general, we should choose the last alternative that know we can display.
+            // RFC 2046 Section 5.1.4 https://datatracker.ietf.org/doc/html/rfc2046#section-5.1.4
+            auto chosen_alternative = alternatives.last_matching([this](auto& alternative) {
+                return is_supported_alternative(alternative);
+            });
+
+            if (alternatives.is_empty()) {
+                GUI::MessageBox::show(window(), String::formatted("Displaying this type of e-mail is currently unsupported."), "Unsupported", GUI::MessageBox::Type::Information);
+                return;
+            }
+            selected_alternative_position = chosen_alternative.value().position;
```

Comment:
You can use `->` instead of `.value().` for some more compact code, but it's not strictly necessary, `.value()` is fine too.